x-pack/plugins/security_solution/server/endpoint/services/actions/clients/crowdstrike/crowdstrike_actions_client.ts

```diff
@@ -0,0 +1,323 @@
+  connectorActions: ActionsClient;
+};
+
+export class CrowdstrikeActionsClient extends ResponseActionsClientImpl {
+  protected readonly agentType: ResponseActionAgentType = 'crowdstrike';
+  private readonly connectorActionsClient: ActionsClient;
+  private readonly getConnector: () => Promise<ConnectorWithExtraFindData>;
+
+  constructor({ connectorActions, ...options }: CrowdstrikeActionsClientOptions) {
+    super(options);
+    this.connectorActionsClient = connectorActions;
+
+    this.getConnector = once(async () => {
+      let connectorList: ConnectorWithExtraFindData[] = [];
+
+      try {
+        connectorList = await this.connectorActionsClient.getAll();
+      } catch (err) {
+        throw new ResponseActionsClientError(
+          `Unable to retrieve list of stack connectors: ${err.message}`,
+          // failure here is likely due to Authz, but because we don't have a good way to determine that,
+          // the `statusCode` below is set to `400` instead of `401`.
+          400,
+          err
+        );
+      }
+      const connector = connectorList.find(({ actionTypeId, isDeprecated, isMissingSecrets }) => {
+        return actionTypeId === CROWDSTRIKE_CONNECTOR_ID && !isDeprecated && !isMissingSecrets;
+      });
+
+      if (!connector) {
+        throw new ResponseActionsClientError(
+          `No Crowdstrike stack connector found`,
+          400,
+          connectorList
+        );
+      }
+
+      this.log.debug(`Using Crowdstrike stack connector: ${connector.name} (${connector.id})`);
+
+      return connector;
+    });
+  }
+
+  protected async writeActionRequestToEndpointIndex(
+    actionRequest: Omit<ResponseActionsClientWriteActionRequestToEndpointIndexOptions, 'hosts'>
+  ): Promise<LogsEndpointAction> {
+    const agentId = actionRequest.endpoint_ids[0];
+    const agentDetails = await this.getAgentDetails(agentId);
+
+    return super.writeActionRequestToEndpointIndex({
+      ...actionRequest,
+      hosts: {
+        [agentId]: { name: agentDetails[0].hostname },
+      },
+    });
+  }
+
+  /**
+   * Sends actions to Crowdstrike directly (via Connector)
+   * @private
+   */
+  private async sendAction(
+    actionType: SUB_ACTION,
+    actionParams: object
+  ): Promise<ActionTypeExecutorResult<unknown>> {
+    const { id: connectorId } = await this.getConnector();
+    const executeOptions: Parameters<typeof this.connectorActionsClient.execute>[0] = {
+      actionId: connectorId,
+      params: {
+        subAction: actionType,
+        subActionParams: actionParams,
+      },
+    };
+
+    this.log.debug(
+      `calling connector actions 'execute()' for Crowdstrike with:\n${stringify(executeOptions)}`
+    );
+
+    const actionSendResponse = await this.connectorActionsClient.execute(executeOptions);
+
+    if (actionSendResponse.status === 'error') {
+      this.log.error(stringify(actionSendResponse));
+
+      // TODO TC: check if we want to handle different errors like 401
```

Comment:
To your question here: is the HTTP status returned from the connector's sub-action?

When I checked the SentinelOne code, I don't remember seen it, thus why I had implemented the S1 client to always return a `500`.

_(Note: I found the errors returned by the S1 sub-actions to be missing a bit of data that would be really beneficial for consumers of that service. I'm not sure if it is a result of how the connector was implemented or the sub-actions framework itself, but I have not had time to look into it. But if the connector implementation has control over the structure of the data returned in an error condition, then I think that `statusCode` and `response.body` would be good to have returned )_

Reply:
So unfortunately I cannot return 401 because that triggers a logout in kibana. I think 500 with a description is good enough. Unless you know a way to return 401 and keep user logged in?

Reply:
ahhh. yes. Since this is a response to our own API, a 401 would not be good. I'm ok with returning the existing `500`